src/tagmanager/tm_workspace.c

```diff
@@ -681,53 +684,336 @@ GPtrArray *tm_workspace_find(const char *name, const char *scope, TMTagType type
 }
 
 
-static void fill_find_tags_array_prefix(GPtrArray *dst, const GPtrArray *src,
-	const char *name, TMParserType lang, guint max_num)
+static gboolean is_valid_autocomplete_tag(TMTag *tag,
+	TMSourceFile *current_file,
+	guint current_line,
+	const gchar *current_scope)
 {
-	TMTag **tag, *last = NULL;
-	guint i, count, num;
+	/* ignore local variables from other files/functions or after current
+	 * line for some languages */
+	return !(tag->type & tm_tag_local_var_t) ||
+		(current_file == tag->file &&
+		 (tm_parser_var_valid_before_declare(tag->lang) || current_line >= tag->line) &&
+		 g_strcmp0(current_scope, tag->scope) == 0);
+}
 
-	if (!src || !dst || !name || !*name)
+
+typedef struct
+{
+	TMSourceFile *file;
+	TMSourceFile *header;
+	guint line;
+	const gchar *scope;
+	TMParserType lang;
+} CopyInfo;
+
+
+static gboolean is_autocomplete_tag(TMTag *tag, CopyInfo *info)
+{
+	return is_valid_autocomplete_tag(tag, info->file, info->line, info->scope) &&
+		tm_parser_langs_compatible(info->lang, tag->lang) &&
+		!tm_tag_is_anon(tag);
+}
+
+static gboolean is_local_tag(TMTag *tag, CopyInfo *info)
+{
+	return (tag->type & tm_tag_local_var_t) &&
+		is_autocomplete_tag(tag, info);
+}
+
+/* tag from current file but not local tag */
+static gboolean is_current_file_tag(TMTag *tag, CopyInfo *info)
+{
+	return tag->file == info->file &&
+		!(tag->type & tm_tag_local_var_t) &&
+		is_autocomplete_tag(tag, info);
+}
+
+/* tag from current file's header but not local tag */
+static gboolean is_header_file_tag(TMTag *tag, CopyInfo *info)
+{
+	return tag->file == info->header &&
+		!(tag->type & tm_tag_local_var_t) &&
+		is_autocomplete_tag(tag, info);
+}
+
+/* any tag except the above categories */
+static gboolean is_workspace_tag(TMTag *tag, CopyInfo *info)
+{
+	return  tag->file != info->file &&
+		tag->file != info->header &&
+		!(tag->type & tm_tag_local_var_t) &&
+		is_autocomplete_tag(tag, info);
+}
+
+
+static guint copy_tags(GPtrArray *dst, TMTag **src, guint src_len, GHashTable *name_table,
+	gint num, gboolean (*predicate) (TMTag *, CopyInfo *), CopyInfo *info)
+{
+	guint i;
+
+	g_return_val_if_fail(src && dst, 0);
+
+	for (i = 0; i < src_len && num > 0; i++)
+	{
+		TMTag *tag = *src;
+		if (predicate(tag, info) &&
```

Comment:
Every predicate function contains the is_autocomplete_tag() check. I think it makes sense to have that call here (before calling the predicate). The function could be suitably named `copy_autocomplete_tags()` 

That should perform a bit better

Reply:
Can do, I would just move it after `predicate()` - `is_autocomplete_tag()` potentially makes string comparison of scopes and it's better to eliminate those with the other conditions first.

Reply:
Hm ok. I was more concerned about calling a function via pointer and hoped it could be avoided. But a string comparison isn't negligible either. And I don't really want to start doing micro-benchmarks here.

Reply:
Yeah, lets leave it and start worrying if it ever actually causes an issue.

Reply:
Call over pointer doesn't matter at all, it's still the same function in the loop and the processor will handle it for free. Having to compare scope for local variables is another story - with local variables enabled we get 3x more tags than before and eliminating the comparison early by checking whether we are in the right file definitely helps.

Reply:
Regardless of the order, IMO the call to `is_autocomplete_tag()` doesn't need to be duplicated in each predicate function but can be called in one place here, especially if the compiler whats to inline it.

Indirect calls aren't as free as direct calls but it lets not focus on that, I don't care much.

Reply:
> Regardless of the order, IMO the call to is_autocomplete_tag() doesn't need to be duplicated in each predicate function but can be called in one place here, especially if the compiler whats to inline it.

Agree.

And if performance is your concern, we can actually do better. I didn't think of it when writing the patch but these calls 
```
		copy_tags(dst, found, count, name_table, max_num - dst->len, is_local_tag, info);
		copy_tags(dst, found, count, name_table, max_num - dst->len, is_current_file_tag, info);
		copy_tags(dst, found, count, name_table, max_num - dst->len, is_header_file_tag, info);
```
could be modified not to perform on the workspace tags but only the tags from the current file or the current header (that is, from `info->file->tags_array` or `info->header->tags_array` (of course after performing tm_tags_find() on them before). That should speed up the search for local tags, current file's tags and header (and possibly include) tags significantly.